### PR TITLE
[SandboxVectorizer] Add MemSeed bundle types

### PR DESCRIPTION
**Pull request by Sterling-Augustine**

### CODE REVIEW

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h [hunk 1]
```diff
@@ -123,5 +123,35 @@ class SeedBundle {
   }
 #endif // NDEBUG
 };
+
+template <typename LoadOrStoreT> class MemSeedBundle : public SeedBundle {
+public:
+  explicit MemSeedBundle(SmallVector<Instruction *> &&SV, ScalarEvolution &SE)
+      : SeedBundle(std::move(SV)) {
+    assert(all_of(Seeds, [](auto *S) { return isa<LoadOrStoreT>(S); }) &&
```
Comment (vporpo) [inline llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h:131]: Perhaps a static assert?
```
static_assert(std::is_same<LoadOrStoreT, LoadInst>::value || std::is_same<LoadOrStoreT, StoreInst>::value, "Expected LoadInst or StoreInst!");
```

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h [hunk 2]
```diff
@@ -123,5 +123,35 @@ class SeedBundle {
...
+      return Utils::atLowerAddress(cast<LoadOrStoreT>(I0),
+                                   cast<LoadOrStoreT>(I1), SE);
+    };
+    std::sort(Seeds.begin(), Seeds.end(), Cmp);
+  }
+  explicit MemSeedBundle(LoadOrStoreT *MemI) : SeedBundle(MemI) {
+    assert(isa<LoadOrStoreT>(MemI) && "Expected Load or Store!");
+  }
+  void insert(sandboxir::Instruction *I, ScalarEvolution &SE) {
+    assert(isa<LoadOrStoreT>(I) && "Expected a Store or a Load!");
```
Comment (vporpo) [inline llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h:143]: Same, static_assert.

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h [hunk 3]
```diff
@@ -123,5 +123,35 @@ class SeedBundle {
   }
 #endif // NDEBUG
 };
+
+template <typename LoadOrStoreT> class MemSeedBundle : public SeedBundle {
```
Comment (vporpo) [inline llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h:127]: A brief description comment?

Comment (Sterling-Augustine) [reply llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h:127]: Done

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h (hunk 1 above)
Comment (Sterling-Augustine) [reply llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h:131]: Done

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h (hunk 2 above)
Comment (Sterling-Augustine) [reply llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h:143]: Done

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h [hunk 4]
```diff
@@ -123,5 +123,44 @@ class SeedBundle {
...
+    assert(all_of(Seeds, [](auto *S) { return isa<LoadOrStoreT>(S); }) &&
+           "Expected Load or Store instructions!");
+    auto Cmp = [&SE](Instruction *I0, Instruction *I1) {
+      return Utils::atLowerAddress(cast<LoadOrStoreT>(I0),
+                                   cast<LoadOrStoreT>(I1), SE);
+    };
+    std::sort(Seeds.begin(), Seeds.end(), Cmp);
+  }
+  explicit MemSeedBundle(LoadOrStoreT *MemI) : SeedBundle(MemI) {
+    static_assert(std::is_same<LoadOrStoreT, LoadInst>::value ||
```
Comment (vporpo) [inline llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h:146]: Hmm actually don't we actually need only one static assert? The compiler should crash if we keep the one in the constructor.

**Review by vporpo — APPROVED** [review]
LG

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h (hunk 4 above)
Comment (Sterling-Augustine) [reply llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SeedCollector.h:146]: There are two constructors. One for a single instruction, one for a list. Not sure which will end up being more used.

**Pull request merged**